File: 240207/빙빙 돌며 숫자 사각형 채우기 2/snail-number-square-2.py
n, m = map(int, input().split())
arr = [[0 for _ in range(m)] for _ in range(n)]

dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]

x, y = 0, 0

arr[x][y] = 1
dir_num = 2

def in_range(x, y):
    return 0 <= x < n and 0 <= y < m

for i in range(2, n*m+1):
    nx, ny = x + dx[dir_num], y + dy[dir_num] 

    if not in_range(nx, ny) or arr[nx][ny] != 0:
        dir_num = (dir_num + 3) % 4
    
    x = x + dx[dir_num]
    y = y + dy[dir_num]
    arr[x][y] = i

for i in range(n):
    for j in range(m):
        print(arr[i][j], end=' ')
    print()

File: 240913/빙빙 돌며 숫자 사각형 채우기 2/snail-number-square-2.js
const fs = require("fs");
let [n, m] = fs.readFileSync(0).toString().trim().split(" ").map(Number);

let arr = Array(n).fill(0).map(() => Array(m).fill(0));
arr[0][0] = 1;
let dx = [1, 0, -1, 0], dy = [0, 1, 0, -1];
let dirNum = 0;
let [x, y] = [0, 0];

for (let i = 2; i <= n * m; i++) {
    let nx = x + dx[dirNum], ny = y + dy[dirNum];

    while (true) {
        if (!inRange(nx, ny) || arr[nx][ny] !== 0) {
            dirNum = (dirNum + 1) % 4;
            nx = x + dx[dirNum], ny = y + dy[dirNum];
        } else {
            break;
        }
    }

    x += dx[dirNum], y += dy[dirNum];
    arr[x][y] = i;
}

for (let i = 0; i < n; i++) {
    console.log(arr[i].join(" "));
}

function inRange(x, y) {
    return x >= 0 && x < n && y >= 0 && y < m;
}